presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddLocalExchanges.java

```diff
@@ -549,7 +549,7 @@ public PlanWithProperties visitTableWriter(TableWriterNode originalTableWriterNo
                 }
             }
             else {
-                PlanWithProperties source = originalTableWriterNode.getSource().accept(this, fixedParallelism());
+                PlanWithProperties source = originalTableWriterNode.getSource().accept(this, defaultParallelism(session));
```

Comment:
What's the difference between `fixedParallelism()` and `defaultParallelism`? 

Also from the commit message

> For partitioned writes the shuffle is added unconditionally.

Curious what does it mean by "partitioned writes"? Do you mean the table write input is fixed distribution? (e.g. following aggregation or join)

Reply:
Just to double check: this is more about avoiding unnecessary local exchange, right?

Reply:
> What's the difference between fixedParallelism() and defaultParallelism?

The difference is subtle. `fixedParallelism()` - FIXED distribution is preferred. `defaultParallelism` - parallel distribution is preferred (either `MULTIPLE` or `FIXED`). Since we are adding a local exchange unconditionally it doesn't make sense to request `FIXED`.

> Curious what does it mean by "partitioned writes"? Do you mean the table write input is fixed distribution? (e.g. following aggregation or join)

`partitioned writes` is when you write into a bucketed table. For bucketed tables a hash based exchange is added unconditionally for correctness. 

> Just to double check: this is more about avoiding unnecessary local exchange, right?

Yeah, since we are adding an exchange on top unconditionally, we don't want the exchange to be added when planning the upstream node